include/bluetooth/services/ots.h

```diff
@@ -493,48 +496,57 @@ struct bt_ots_cb {
 	 *  This callback is called whenever a new object is created.
 	 *  Application can reject this request by returning an error
 	 *  when it does not have necessary resources to hold this new
-	 *  object. This callback is also triggered when the server
-	 *  creates a new object with bt_ots_obj_add() API.
+	 *  object or it does not support the requested object type.
+	 *  This callback is also triggered when the server creates a
+	 *  new object with bt_ots_obj_add() API.
 	 *
-	 *  @param ots  OTS instance.
-	 *  @param conn The connection that is requesting object creation or
-	 *              NULL if object is created by the following function:
-	 *              bt_ots_obj_add().
-	 *  @param id   Object ID.
-	 *  @param init Object initialization metadata.
+	 *  @param ots       OTS instance.
+	 *  @param conn      The connection that is requesting object creation or
+	 *                   NULL if object is created by the following function:
+	 *                   bt_ots_obj_add().
+	 *  @param id        Object ID.
+	 *  @param user_data In: NULL when called by object creation procedure or
+	 *                       user data passed in when called by bt_ots_obj_add() API.
+	 *                   Out: User data to be associated with this object.
+	 *  @param init      Object initialization metadata.
 	 *
 	 *  @return 0 in case of success or negative value in case of error.
 	 *  Possible return values:
+	 *  -ENOTSUP if object type is not supported
 	 *  -ENOMEM if no available space for new object.
 	 */
 	int (*obj_created)(struct bt_ots *ots, struct bt_conn *conn,
-			   uint64_t id,
-			   const struct bt_ots_obj_metadata *init);
+			   uint64_t id, void **user_data,
```

Comment:
The solution you have suggested solves the problem of passing information to the `ots_obj_created` by the caller of `bt_ots_obj_add`. It does not address the issue of easily looking up information associated with an object.

The association problem can be passed on to the application by keeping a map of object IDs to the user data, adding the `user_data` field just seemed like a good extension to the object manager that is already keeping a map of object IDs to some object data.

I was not aware of the desire to share the APIs between the server and the client that does shine a new light on the API decisions. I will make changes to remove the user data and use a local application mapping.

Reply:
@ekohandel  The shared API is still a topic we are discussing for some of the other services, so it may be that we want to have fully individual APIs for the server and client.